= Mathematical Principles of Graph Neural Networks and Geometric Deep Learning for Bioinformatics
Konstantinos Krampis, Eric Ross, Olorunseun O. Ogunwobi, Grace Ma, Raja Mazumder, Claudia Wultsch
:stem:
:bibtex-file: GDL-proto.bib

== ABSTRACT
 Following the exponential growth of the machine and deep learning
field in recent years, artificial neural network models have found significant
applications towards bioinformatics data analysis. The various omics datasets
are commonly represented in a graph format, such as for example protein or gene
interaction networks, molecular structures and cellular signalling pathways.
Unlike structured images, video and text that are commonly used for training
artificial neural networks, graph data are in the non-euclidean domain and
require significantly different algorithmic approaches. The machine learning
research community, has developed a range of new algorithms for training
artificial neural networks with graph data. These novel approaches and their
importance for the bioinformatics field is established herein, through
exhibition of the undelying mathematical foundations from group theory,
functional analysis and linear algebra. Furthermore, it is argued that the most
recent developments in the field such as geometric deep learning on data
manifolds, can also significantly accelerate scientific discovery in
bioinformatics by enabling new approaches to understand complex datasets.
Finally, we conclude this opinion article with the options for implementations
through the available software frameworks, as guideline for transitioning from
the mathematical principles to practical graph machine learning tools for
bioinformatics applications.


== INTRODUCTION

Symmetry and invariance is a central concept in physical, mathematical and
biological systems, in addition to other areas of human endeavor such as social
networks. It has been established since the early 20th century that
fundamental laws of nature originate in symmetry, for example through the work
of Noether and Wigner cite:[noether1918invariante].In the last decade, technologies such as genomic
sequencing have enabled an exponential increase [REF] of the data that describe
the fundamental elements, structure and function of biological systems. Other
endeavors from fields as diverse as physics [REF collider], and the rise of
social media platforms [ref], have resulted in datasets of scale not previously
available to scientists. This data explosion, has also been fundamental for the
ever accelerating advancements in the field of machine learning, deep learning
and artificial intelligence, where we now  have algorithms that can be trained
to make discoveries from the data at a level that matches closely human
intuition.

As in any field (including bioinformatics) that develops rapidly whithin the
span of a few years,deep learning and artificial intelligence researchers have
developed hundreds of successful algorithms, however with a few unifying
principles . In a seminal `proto-book` by Bronstein et al. cite:[bronstein2021geometric], 
a range of systematization principles for the different artificial
neural network architectures from the deep learning field was presented, based
on the concepts of symmetry that is formalized within the mathematical field of
group theory. The authors also introduced the concept of Geometric Deep
Learning, and demonstrated how the group theoretic principles of iso- and
auto-morphism, along with invariance and equivariance of functions, can be used
in composing and describing various deep learning algorithms. 

== THE STRUCTURE AND FUNCTION OF ARTIFICIAL INTELLIGENCE

Before proceeding towards introducing these concepts further, we will describe
the structure and function of deep learning and artificial neural networks that
are the foundation of artificial intelligence,through a simple mathematical
description. Assume a dataset consisting of _n_ pairs of stem:[{x_i,y_i}_j^n]
,with the stem:[x_i] being the data points and stem:[y_i] their labels. Each
data point can be for example be a set of numbers, vectors (arrays of numbers), 
storing the data for image pixels, or graph data structures 
composed by nodes and edges representing drugs or other chemical molecules. 
Whether the data originate from lab experiments, clinical researh, or public databases,
as a graph vector stem:[x_i] for example representing a molecule, could have a
binary (two-option) label of stem:[y_i=1] "inhibits cancer growth", or stem:[y_i=0] "does not
inhibit cancer". The labels can also be continuous numbers such as for example
stem:[y_i=0.3] meaning 30% inhibition, while also each stem:[y_i] could be a set
such as stem:[y_i = {0,1,0}] representing terms for a molecule such as 'no inhibition',
'yes for toxicity', 'not metabolized', with 'not' for 0 and 'yes' 1.

Similar to a brain neural network, the computational abstractions used in machine learning 
and artificial intelligence, model neurons as signal aggregators and thresholding units. 
Specifically, each artificial neuron performs a summation of incoming signals from its connected 
neurons, and "fires" when the aggregated signals reach a certain threshold. This can be described with
the summation stem:[sigma_(kl)=sum_(k=1)^n w_(kl)**x_(kl) - b], where by following the
biological analogy the stem:[w_(kl)] represents the connection weight, as learning takes place in the
biological brain by strengthening connections among neurons. The subscripts represent respectively
a connection from dendrite _l_ of neighboring neuron _k_ (REF), and similarly stem:[x_(kl)]
is the incoming signal. The term _b_ represents a threshold that needs to be surpassed with 
stem:[w_(kl)**x_(kl)>b] in order for the neuron to activate and "fire" a signal. In the simplest 
version of "fully connected" artificial neural networks, each neuron _k_ in the network receives the 
input _i_, resulting in _k*i_ incoming and outgoing connections per neuron.  The activation of the neuron 
is also determined by an additional thresholding or otherwise "logit" represented as stem:[varphi(sigma_(kl))].
which is usually non-linear (REF). Following the summation of incoming signals, passing of the threshold
and computation of the activation function, the final value which is the output of the firing neuron,
is transmitted to its connected neurons through the outgoing connections or "axons" in the biological analogy. 
Multiple layers of such computational models of neurons connected together, along with multiple connections 
per layer each having each own weight stem:[w_(kl)], forms an Artificial Neural Network (ANN).

Mathematically, a trained ANN is a function stem:[f] that predicts the labels stem:[y_pred_i] for the input 
data stem:[x_i], such as for example 'no inhibition', 'yes for toxicity' etc. for graph data representing  
drug molecules.  The function stem:[f] is non-linear and is estimated by fitting a training dataset, which 
correlates labels stem:[y_i] to data points stem:[x_i]. Most important, stem:[f] is a composition
of the stem:[varphi(sigma_(kl))] thresholding functions of the each neuron interconnected in
the artificial network graph cite:[bronstein2021geometric]. This is key aspect of the ANNs,
since the non-linearities present in each individual logit neuron function stem:[sigma_(kl)], 
are aggregated across the neural network and on the function stem:[f], which enables fitting complex 
multi-dimensional input data with non-linear distributions. This is the key fact that enables ANNs 
to achieve higher clasification power compared to traditional statistical 
methods (REF). With hundreds of papers and monographs that have been written on the
technical details of training ANNs,and since the focus of the present review
manuscript are networks designed with graph and other non-euclidean datasets, we
will next attempt to summarize the training in few sentences and refer the reader to
REF for further details. 

In summary, the ANN algorithm is using the training data to identify a function stem:[f] that
predicts labels from the data such as stem:[y_(pred_i)=f(x_i)]. As mentioned previously,
stem:[f] is a composition of the stem:[varphi(sigma_(kl))] functions of the each neuron in the ANN,
and as such the training is the estimatiion of the weights stem:[w_kl] of the neuron connections, 
while minimizing the error stem:[E] based on the difference between stem:[y_(pred_i)] and stem:[y_i].
This error quantifies the neural network precision by comparing 
predicted stem:[y_(pred_i)] and actual labels stem:[y_i]. The neuron connection weight stem:[w_kl] 
estimation by the algorithm takes place through fitting the network function stem:[f] on a large 
training dataset of stem:[{x_i,y_i}_i^n], pairs of input data and labels, while it evaluates the error 
stem:[E] using smaller testing and validation data subsets. The training algorithm works by making 
an initial estimated guess for initializing the weights, and then performing multiple cycles (called "epochs")
of fitting stem:[x_i] the training data to the network (REF). At the end of each cycle "backpropagation" is
performed [REF], which involves gradient descent optimization, in order to fine tune the weights of the 
individual neurons in stem:[sigma_(kl)=sum_(k=1)^n w_(kl)**x_(kl) + b]. 
The gradient descent (REF) searches the possible combinations of weight values, and since it is a heuristic
algorithm it minimizes but cannot reach zero error stem:[E]. At the completion of multiple training cycles 
the training algorithm identifies a set of weights which best fit the data model, and the ANN settles on 
the optimal parameters that estimate the stem:[varphi(sigma_(kl))] function for each interconnected neuron. 
Consequently, the overall stem:[f(x_i)] is also estimated,since it is the composition 
of the individual stem:[varphi(sigma_(kl))] neuron functions.  Once the artificial neural network training 
has ben completed by finding the most optimal set of weights, it is now ready
to be used for label prediction with new, unknown stem:[x_i] data.

== ARTIFICIAL INTELLIGENCE, GROUP THEORY, SYMMETRY AND INVARIANCE

For the remaining of this manuscript, we will discuss the principles of symmetry and
invariance through the lens of group theory, and how these related to the discerning
power of the trained neural network in relation to variance in the input data. We will
again follow the terminology of Bronstein et al. (REF), where we define the input data
stem:[x_i] to be from a domain stem:[Omega], which is defined accordingly based on the
type of data. For example, with image data being 2-dimensional grids of _n x n _ pixels 
(each pixel having an integer value for each intensity), we would then define 
stem:[Omega: ZZ_n xx ZZ_n] as integer grid of images. More specifically, if the input
data are for 3-layer color images (green, blue red) stem:[x_i:Omega to ZZ^3] of cancer tissue,
we would then define a signal as stem:[X(Omega)] and the overall neural network estimator function
operating on this signal as stem:[f("X"(Omega))]. Having defined these essential
terms, we will proceed towards discussing the concepts of symmetry and invariance through
the lens of group theory, in order to examine the classifying power of ANNs in relation
to variance in the data.

In a broader definition, symmetry is the study of space and structure, whether this refers
to geometrical figures and algebraic structures in mathematics, particle configurations in
physics, or organism structures in biology. Invariance of an object under transformation, is
the property of changing the position of the object in space, such as rotating a drug molecule
or shifting a cancer histology image, leaving the properties of the object unchanged (i.e. the
drug remains potent and the tissue is recognized as cancerous from the image). When artificial 
neural networks act as function estimators stem:[f("X"(Omega) to "Y")] to predict output 
labels (i.e stem:[mathcal "Y"] = potent drug / not potent, 


bibliography::[]

